### PR TITLE
 [Improvement]: Refactor ApiError and ErrorKind to use thiserror #1601

### LINKED ISSUE #1601: [Improvement]: Refactor ApiError and ErrorKind to use thiserror
### Describe the improvement

https://github.com/shuttle-hq/shuttle/blob/c5ddea00829ff2a3c75676c97a52ef684e12a255/common/src/models/error.rs#L8-L12

https://github.com/shuttle-hq/shuttle/blob/c5ddea00829ff2a3c75676c97a52ef684e12a255/common/src/models/error.rs#L33-L34

These two structs in shuttle-common currently use manual implementation for error message formatting.
The `impl From<ErrorKind> for ApiError` block has lots of repetition.

The idea is to use `thiserror`'s template string attributes that automate string formatting for the inner enum type. That way, the big match statement can be used for just selecting the status code.

All error messages and codes should remain the same after this change, unless you find some reasonable improvements to make.

---

<sub>This issue is open to contribution. Comment if you want to work on it. Feel free to ask for help and guidance here or in our Discord.</sub>

### Duplicate declaration

- [X] I have searched the issues and this improvement has not been requested before.

### PR DESCRIPTION
**Pull request by biplab5464**
## Description of change
use thiserror's template string attributes that automate string formatting for the inner enum type
```
ErrorKind::ProjectHasResources(resources) => {
                let resources = resources.join(", ");
                return Self {
                    message: format!("Could not automatically delete the following resources: {}. Please reach out to Shuttle support for help.", resources),
                    status_code: StatusCode::INTERNAL_SERVER_ERROR.as_u16(),
                };
            }
``` 
i didn't change this part because Vec<String> is change to `[hello,world]`

Closes #1601

### CODE REVIEW

File: common/src/models/error.rs [hunk 1]
```diff
@@ -30,114 +31,107 @@ impl Display for ApiError {
 
 impl std::error::Error for ApiError {}
```
Comment (jonaro00) [inline common/src/models/error.rs:32]: This might not be needed if thiserror::Error is derived.

File: common/src/models/error.rs [hunk 2]
```diff
@@ -30,114 +31,107 @@ impl Display for ApiError {
...
+    #[error("Request is missing a key")]
     KeyMissing,
+    #[error("The 'Host' header is invalid")]
     BadHost,
+    #[error("Request has an invalid key")]
     KeyMalformed,
+    #[error("Request is missing a key")]
     Unauthorized,
+    #[error["You cannot create more projects. Delete some projects first."]]
     Forbidden,
```
Comment (jonaro00) [inline common/src/models/error.rs:45]: These are not the same as before. Was this intentional?

File: common/src/models/error.rs [hunk 3]
```diff
@@ -30,114 +31,107 @@ impl Display for ApiError {
...
+            ErrorKind::ProjectUnavailable => StatusCode::BAD_GATEWAY,
+            ErrorKind::TooManyProjects => StatusCode::FORBIDDEN,
+            ErrorKind::ProjectHasRunningDeployment => StatusCode::INTERNAL_SERVER_ERROR,
+            ErrorKind::ProjectHasBuildingDeployment => StatusCode::BAD_REQUEST,
+            ErrorKind::ProjectCorrupted => StatusCode::BAD_REQUEST,
             ErrorKind::ProjectHasResources(resources) => {
                 let resources = resources.join(", ");
                 return Self {
                     message: format!("Could not automatically delete the following resources: {}. Please reach out to Shuttle support for help.", resources),
                     status_code: StatusCode::INTERNAL_SERVER_ERROR.as_u16(),
```
Comment (jonaro00) [inline common/src/models/error.rs:117]: I think the debug formatting in the error macro is good enough, so this shouldn't be needed.

File: common/src/models/error.rs [hunk 4]
```diff
@@ -30,114 +31,107 @@ impl Display for ApiError {
...
+            ErrorKind::ProjectAlreadyExists => StatusCode::BAD_REQUEST,
+            ErrorKind::OwnProjectAlreadyExists(_) => StatusCode::BAD_REQUEST,
+            ErrorKind::InvalidCustomDomain => StatusCode::BAD_REQUEST,
+            ErrorKind::CustomDomainNotFound => StatusCode::NOT_FOUND,
+            ErrorKind::CustomDomainAlreadyExists => StatusCode::BAD_REQUEST,
+            ErrorKind::Unauthorized => StatusCode::UNAUTHORIZED,
+            ErrorKind::Forbidden => StatusCode::FORBIDDEN,
+            ErrorKind::NotReady => StatusCode::INTERNAL_SERVER_ERROR,
+            ErrorKind::DeleteProjectFailed => StatusCode::INTERNAL_SERVER_ERROR,
+            ErrorKind::CapacityLimit => StatusCode::SERVICE_UNAVAILABLE,
```
Comment (jonaro00) [inline common/src/models/error.rs:131]: It is possible to combine multiple arms with  `A | B | C ... => StatusCode:: ...`, but I'm not sure if that makes it more or less readable.

**Review by jonaro00 — COMMENTED** [review]
Looks pretty good overall!

File: common/src/models/error.rs (hunk 2 above)
Comment (jonaro00) [reply common/src/models/error.rs:45]: The 2nd one also has brackets for some reason :sweat_smile:

**Comment by biplab5464** [comment]
hey it is midnight here, so i have some changes but have some more changes 
i will do it tomorrow

File: common/src/models/error.rs [hunk 5]
```diff
@@ -30,114 +31,101 @@ impl Display for ApiError {
...
+            ErrorKind::Internal => StatusCode::INTERNAL_SERVER_ERROR,
+            ErrorKind::KeyMissing => StatusCode::UNAUTHORIZED,
+            ErrorKind::ServiceUnavailable => StatusCode::SERVICE_UNAVAILABLE,
+            ErrorKind::KeyMalformed => StatusCode::BAD_REQUEST,
+            ErrorKind::BadHost => StatusCode::BAD_REQUEST,
+            ErrorKind::UserNotFound => StatusCode::NOT_FOUND,
+            ErrorKind::UserAlreadyExists => StatusCode::BAD_REQUEST,
+            ErrorKind::ProjectNotFound(_) => StatusCode::NOT_FOUND,
+            ErrorKind::ProjectNotReady => StatusCode::SERVICE_UNAVAILABLE,
+            // "not ready" is matched against in cargo-shuttle for giving further instructions on project deletion
```
Comment (jonaro00) [inline common/src/models/error.rs:107]: Could you move this comment up to the ProjectNotReady error message?

Comment (jonaro00) [reply common/src/models/error.rs:106]: Sorry, by error message I meant the corresponding `#[error()]` annotation in the enum definition

**Comment by biplab5464** [comment]
is my pr is ready for merged

**Review by jonaro00 — APPROVED** [review]

**Review by chesedo — APPROVED** [review]
LGTM

**Comment by biplab5464** [comment]
Thanks for your cooperation : )

**Pull request merged**